contracts/bridges/FxGovernorTunnel.sol

```diff
@@ -0,0 +1,147 @@
+        // This is possible only if all the checks in the message process function pass and the contract calls itself
+        if (msg.sender != address(this)) {
+            revert SelfCallOnly(msg.sender, address(this));
+        }
+
+        // Check for the zero address
+        if (newRootGovernor == address(0)) {
+            revert ZeroAddress();
+        }
+
+        rootGovernor = newRootGovernor;
+    }
+
+    /// @dev Process message received from the Root Tunnel.
+    /// @notice This is called by onStateReceive function. The sender must be the Root Governor address (Timelock).
+    /// @param stateId Unique state id.
+    /// @param rootMessageSender Root message sender.
+    /// @param data Bytes message sent from the Root Tunnel. The data must be encoded as a set of continuous
+    ///        transactions packed into a single buffer, where each transaction is composed as follows:
+    ///        - target address of 20 bytes (160 bits)
+    ///        - value of 12 bytes (96 bits), as a limit for all of Autonolas ecosystem contracts
+    ///        - payload length of 4 bytes (32 bits), as 2^32 - 1 characters is more than enough to fill a whole block
+    ///        - payload as bytes, with the length equal to the specified payload length
+    function processMessageFromRoot(uint256 stateId, address rootMessageSender, bytes memory data) external override {
+        // Check for the Fx Child address
+        if(msg.sender != fxChild) {
+            revert FxChildOnly(msg.sender, fxChild);
+        }
+
+        // Check for the Root Governor address
+        if(rootMessageSender != rootGovernor) {
+            revert RootGovernorOnly(rootMessageSender, rootGovernor);
+        }
+
+        // Check for the correct data length
+        uint256 dataLength = data.length;
+        if (dataLength < DEFAULT_DATA_LENGTH) {
+            revert IncorrectDataLength(DEFAULT_DATA_LENGTH, data.length);
+        }
+        
+        // Unpack and process the data
+        for (uint256 i = 0; i < dataLength;) {
+            address target;
+            uint96 value;
+            uint32 payloadLength;
+            // solhint-disable-next-line no-inline-assembly
+            assembly {
+                // First 20 bytes is the address (160 bits)
+                i := add(i, 20)
+                target := mload(add(data, i))
+                // Offset the data by 12 bytes of value (96 bits)
+                i := add(i, 12)
+                value := mload(add(data, i))
+                // Offset the data by 4 bytes of payload length (32 bits)
+                i := add(i, 4)
+                payloadLength := mload(add(data, i))
+            }
+
+            // Get the payload
+            bytes memory payload = new bytes(payloadLength);
+            for (uint256 j = 0; j < payloadLength; ++j) {
+                payload[j] = data[i + j];
+            }
+            // Offset the data by the payload number of bytes
+            i += payloadLength;
```

Comment:
Get the `payload` itself according to its length, and increase the amount of processed bytes.